### PR TITLE
Allow any extension for the generated JavaScript

### PR DESCRIPTION
**Pull request by jchavarri**
There's no reason why only a subset of file extensions is allowed for the generated JavaScript files. This PR allows any extension, no validation is performed as this is done in a previous step.

We might want to consider "massaging" this PR by rethinking `Ext_js_suffix` (e.g. making  `t` abstract), as I don't think it makes a lot of sense to keep the variant type. But I wanted to open the PR for early feedback.

### CODE REVIEW

**Review by anmonteiro — COMMENTED** [review]
Agreed on making it abstract. Suggestion:

- make `Ext_js_suffix` abstract
- validate extension starts with `.`, ends with not-`.`
- make `type t = string` in the implementation


what do you think?

**Comment by jchavarri** [comment]
Sounds good.

**Comment by jchavarri** [comment]
@anmonteiro I applied the changes we discussed.

File: jscomp/ext/ext_js_suffix.ml [hunk 1]
```diff
@@ -1,19 +1,24 @@
...
-  match () with
-  | () when x = Literals.suffix_js -> Js
-  | () when x = Literals.suffix_bs_js -> Bs_js
-  | () when x = Literals.suffix_mjs -> Mjs
-  | () when x = Literals.suffix_cjs -> Cjs
-  | _ -> Unknown_extension
+  match String.length x with
+  | 0 -> raise (Invalid_argument "File extension can not be empty")
+  | _ -> (
+      let first = String.get x 0 in
```
Comment (anmonteiro) [inline jscomp/ext/ext_js_suffix.ml:9]: we know it's at least length == 1
```suggestion
      let first = String.unsafe_get x 0 in
```

File: jscomp/ext/ext_js_suffix.ml [hunk 2]
```diff
@@ -1,19 +1,24 @@
...
+      match first with
+      | '.' -> (
+          let last = String.get x (String.length x - 1) in
+          match last with
+          | '.' ->
+              raise
+                (Invalid_argument
+                   (Printf.sprintf "File extension %s does not end with '.'" x))
```
Comment (anmonteiro) [inline jscomp/ext/ext_js_suffix.ml:17]:
```suggestion
  | length -> (
      let first = String.unsafe_get x 0 in
      match first with
      | '.' -> (
          match String.unsafe_get x (length - 1) with
          | '.' ->
              raise
                (Invalid_argument
                   (Printf.sprintf "File extension %s cannot end with '.'" x))
```

File: test/blackbox-tests/file-extensions.t [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  
+  console.log("hello");
+  
+  /*  Not a pure module */
+
+  $ rm *
+  $ cat > x.ml <<EOF
+  > let () = Js.log "hello"
+  > EOF
+  $ melc -impl x.ml -o x.foo.
```
Comment (anmonteiro) [inline test/blackbox-tests/file-extensions.t:56]: this is weird. we should check that the extension doesn't start / end with `.`.

**Comment by jchavarri** [comment]
Why using `unsafe_get`? this function does not have any comments / docs and the only thing I found says "The following is for system use only. Do not call directly.".

File: test/blackbox-tests/file-extensions.t (hunk 3 above)
Comment (jchavarri) [reply test/blackbox-tests/file-extensions.t:56]: There is some function `Ext_filename.get_all_extensions_maybe` that runs before the `Ext_js_suffix` functions has a chance to kick in:

https://github.com/melange-re/melange/blob/dc6723376ae99212440b98d8007162eb4db3ce9a/bin/melc.ml#L309

Should we get rid of it or adapt it to ignore extensions ending with `.`?

**Comment by anmonteiro** [comment]
> Why using `unsafe_get`? this function does not have any comments / docs and the only thing I found says "The following is for system use only. Do not call directly.".

`unsafe_get` doesn't do bounds checking, which we're already doing in the `match`.

File: test/blackbox-tests/file-extensions.t (hunk 3 above)
Comment (anmonteiro) [reply test/blackbox-tests/file-extensions.t:56]: probably a good idea to adapt, perhaps in a follow-up PR

**Review by anmonteiro — APPROVED** [review]

**Pull request merged**